File: distinct_subsequences.ex
defmodule DistinctSubsequences do
  @doc """
  Counts the number of distinct subsequences of s which equals t.
  Uses 2D dynamic programming approach.
  
  ## Examples
      
      iex> DistinctSubsequences.num_distinct("rabbbit", "rabbit")
      3
      
      iex> DistinctSubsequences.num_distinct("babgbag", "bag") 
      5
  """
  def num_distinct(s, t) do
    s_chars = String.to_charlist(s)
    t_chars = String.to_charlist(t)
    m = length(s_chars)
    n = length(t_chars)
    
    # Initialize DP table with 0s
    dp = for _ <- 0..m, do: for _ <- 0..n, do: 0
    
    # Base case: empty string t can be formed in 1 way from any string s
    dp = for i <- 0..m, reduce: dp do
      acc -> List.replace_at(acc, i, List.replace_at(Enum.at(acc, i), 0, 1))
    end
    
    # Fill the DP table
    fill_dp(dp, s_chars, t_chars, 1, 1, m, n)
  end
  
  # Space optimized version using 1D array
  def num_distinct_optimized(s, t) do
    s_chars = String.to_charlist(s)
    t_chars = String.to_charlist(t)
    m = length(s_chars)
    n = length(t_chars)
    
    # Initialize 1D DP array
    dp = [1] ++ for _ <- 1..n, do: 0
    
    # Process each character in s
    Enum.with_index(s_chars)
    |> Enum.reduce(dp, fn {s_char, i}, acc ->
      # Process from right to left to avoid overwriting values we need
      Enum.reduce(n..1, acc, fn j, dp_acc ->
        t_char = Enum.at(t_chars, j - 1)
        if s_char == t_char do
          current_val = Enum.at(dp_acc, j)
          prev_val = Enum.at(dp_acc, j - 1)
          List.replace_at(dp_acc, j, current_val + prev_val)
        else
          dp_acc
        end
      end)
    end)
    |> Enum.at(n)
  end
  
  # Recursive solution with memoization
  def num_distinct_recursive(s, t) do
    s_chars = String.to_charlist(s)
    t_chars = String.to_charlist(t)
    memo = %{}
    
    {result, _} = helper(s_chars, t_chars, 0, 0, memo)
    result
  end
  
  # Helper function for filling DP table
  defp fill_dp(dp, s_chars, t_chars, i, j, m, n) when i > m do
    # Return the final result
    Enum.at(Enum.at(dp, m), n)
  end
  
  defp fill_dp(dp, s_chars, t_chars, i, j, m, n) when j > n do
    # Move to next row
    fill_dp(dp, s_chars, t_chars, i + 1, 1, m, n)
  end
  
  defp fill_dp(dp, s_chars, t_chars, i, j, m, n) do
    s_char = Enum.at(s_chars, i - 1)
    t_char = Enum.at(t_chars, j - 1)
    
    # Always have the option to skip current character in s
    skip_val = Enum.at(Enum.at(dp, i - 1), j)
    
    new_val = if s_char == t_char do
      # If characters match, we can also use the current character
      use_val = Enum.at(Enum.at(dp, i - 1), j - 1)
      skip_val + use_val
    else
      skip_val
    end
    
    # Update DP table
    new_row = List.replace_at(Enum.at(dp, i), j, new_val)
    new_dp = List.replace_at(dp, i, new_row)
    
    # Continue to next column
    fill_dp(new_dp, s_chars, t_chars, i, j + 1, m, n)
  end
  
  # Helper function for recursive solution
  defp helper(s_chars, t_chars, i, j, memo) do
    cond do
      # Base cases
      j == length(t_chars) -> {1, memo}  # Found a valid subsequence
      i == length(s_chars) -> {0, memo}  # No more characters in s
      
      # Check memoization
      Map.has_key?(memo, {i, j}) -> {Map.get(memo, {i, j}), memo}
      
      true ->
        # Always have the option to skip current character in s
        {skip_result, memo1} = helper(s_chars, t_chars, i + 1, j, memo)
        
        result = if Enum.at(s_chars, i) == Enum.at(t_chars, j) do
          # If characters match, we can also use the current character
          {use_result, memo2} = helper(s_chars, t_chars, i + 1, j + 1, memo1)
          skip_result + use_result
        else
          skip_result
        end
        
        new_memo = Map.put(memo1, {i, j}, result)
        {result, new_memo}
    end
  end
  
  # Test function
  def run_tests do
    IO.puts("Testing Distinct Subsequences Solutions:")
    IO.puts("")
    
    # Test case 1
    s1 = "rabbbit"
    t1 = "rabbit"
    IO.puts("Test 1:")
    IO.puts("s = \"#{s1}\", t = \"#{t1}\"")
    IO.puts("2D DP: #{num_distinct(s1, t1)}")
    IO.puts("1D DP: #{num_distinct_optimized(s1, t1)}")
    IO.puts("Recursive: #{num_distinct_recursive(s1, t1)}")
    IO.puts("Expected: 3")
    IO.puts("")
    
    # Test case 2
    s2 = "babgbag"
    t2 = "bag"
    IO.puts("Test 2:")
    IO.puts("s = \"#{s2}\", t = \"#{t2}\"")
    IO.puts("2D DP: #{num_distinct(s2, t2)}")
    IO.puts("1D DP: #{num_distinct_optimized(s2, t2)}")
    IO.puts("Recursive: #{num_distinct_recursive(s2, t2)}")
    IO.puts("Expected: 5")
    IO.puts("")
    
    # Test case 3
    s3 = "aaa"
    t3 = "aa"
    IO.puts("Test 3:")
    IO.puts("s = \"#{s3}\", t = \"#{t3}\"")
    IO.puts("2D DP: #{num_distinct(s3, t3)}")
    IO.puts("1D DP: #{num_distinct_optimized(s3, t3)}")
    IO.puts("Recursive: #{num_distinct_recursive(s3, t3)}")
    IO.puts("Expected: 3")
    IO.puts("")
    
    # Edge case: empty t
    s4 = "abc"
    t4 = ""
    IO.puts("Test 4:")
    IO.puts("s = \"#{s4}\", t = \"#{t4}\"")
    IO.puts("2D DP: #{num_distinct(s4, t4)}")
    IO.puts("1D DP: #{num_distinct_optimized(s4, t4)}")
    IO.puts("Recursive: #{num_distinct_recursive(s4, t4)}")
    IO.puts("Expected: 1")
  end
end

# Run the tests
DistinctSubsequences.run_tests()

File: distinct_subsequences.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    // 2D DP Solution - O(m*n) space
    int numDistinct2D(string s, string t) {
        int m = s.length();
        int n = t.length();
        
        // dp[i][j] = number of ways to form t[0...j-1] using s[0...i-1]
        vector<vector<long long>> dp(m + 1, vector<long long>(n + 1, 0));
        
        // Base case: empty string t can be formed in 1 way from any string s
        for (int i = 0; i <= m; i++) {
            dp[i][0] = 1;
        }
        
        // Fill the DP table
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                // Always have the option to skip current character in s
                dp[i][j] = dp[i-1][j];
                
                // If characters match, we can also use the current character
                if (s[i-1] == t[j-1]) {
                    dp[i][j] += dp[i-1][j-1];
                }
            }
        }
        
        return (int)dp[m][n];
    }
    
    // Space Optimized Solution - O(n) space
    int numDistinct(string s, string t) {
        int m = s.length();
        int n = t.length();
        
        // Use 1D array to save space
        vector<long long> dp(n + 1, 0);
        dp[0] = 1; // Base case: empty string can be formed in 1 way
        
        for (int i = 1; i <= m; i++) {
            // Process from right to left to avoid overwriting values we need
            for (int j = n; j >= 1; j--) {
                if (s[i-1] == t[j-1]) {
                    dp[j] += dp[j-1];
                }
                // If characters don't match, dp[j] remains the same
            }
        }
        
        return (int)dp[n];
    }
    
    // Recursive solution with memoization for better understanding
    int numDistinctRecursive(string s, string t) {
        vector<vector<int>> memo(s.length() + 1, vector<int>(t.length() + 1, -1));
        return helper(s, t, 0, 0, memo);
    }
    
private:
    int helper(const string& s, const string& t, int i, int j, vector<vector<int>>& memo) {
        // Base cases
        if (j == t.length()) return 1; // Found a valid subsequence
        if (i == s.length()) return 0; // No more characters in s
        
        if (memo[i][j] != -1) return memo[i][j];
        
        // Always have the option to skip current character in s
        int result = helper(s, t, i + 1, j, memo);
        
        // If characters match, we can also use the current character
        if (s[i] == t[j]) {
            result += helper(s, t, i + 1, j + 1, memo);
        }
        
        return memo[i][j] = result;
    }
};

int main() {
    Solution sol;
    
    // Test case 1
    string s1 = "rabbbit";
    string t1 = "rabbit";
    cout << "Test 1:" << endl;
    cout << "s = \"" << s1 << "\", t = \"" << t1 << "\"" << endl;
    cout << "2D DP: " << sol.numDistinct2D(s1, t1) << endl;
    cout << "1D DP: " << sol.numDistinct(s1, t1) << endl;
    cout << "Recursive: " << sol.numDistinctRecursive(s1, t1) << endl;
    cout << "Expected: 3" << endl << endl;
    
    // Test case 2
    string s2 = "babgbag";
    string t2 = "bag";
    cout << "Test 2:" << endl;
    cout << "s = \"" << s2 << "\", t = \"" << t2 << "\"" << endl;
    cout << "2D DP: " << sol.numDistinct2D(s2, t2) << endl;
    cout << "1D DP: " << sol.numDistinct(s2, t2) << endl;
    cout << "Recursive: " << sol.numDistinctRecursive(s2, t2) << endl;
    cout << "Expected: 5" << endl << endl;
    
    // Additional test case
    string s3 = "aaa";
    string t3 = "aa";
    cout << "Test 3:" << endl;
    cout << "s = \"" << s3 << "\", t = \"" << t3 << "\"" << endl;
    cout << "2D DP: " << sol.numDistinct2D(s3, t3) << endl;
    cout << "1D DP: " << sol.numDistinct(s3, t3) << endl;
    cout << "Recursive: " << sol.numDistinctRecursive(s3, t3) << endl;
    cout << "Expected: 3" << endl << endl;
    
    // Edge case: empty t
    string s4 = "abc";
    string t4 = "";
    cout << "Test 4:" << endl;
    cout << "s = \"" << s4 << "\", t = \"" << t4 << "\"" << endl;
    cout << "2D DP: " << sol.numDistinct2D(s4, t4) << endl;
    cout << "1D DP: " << sol.numDistinct(s4, t4) << endl;
    cout << "Recursive: " << sol.numDistinctRecursive(s4, t4) << endl;
    cout << "Expected: 1" << endl;
    
    return 0;
}